compiler/annotation/match_inference.rs

```diff
@@ -281,6 +282,9 @@ impl TypeInferenceGraph<'_> {
         for nested_graph in &mut self.nested_disjunctions {
             nested_graph.prune_self_from_vertices(&self.vertices)
         }
+        for nested_graph in &mut self.nested_optionals {
+            nested_graph.prune_constraints_from_vertices()
+        }
```

Comment:
Optionals shouldn't be involved here since they can't affect the annotations of the parent.
They should be the in the (aptly named) `prune_types_for_nested_negations_and_optionals`